src/session.rs

```diff
@@ -350,3 +366,169 @@ where
+    use http::{
+        header::{COOKIE, SET_COOKIE},
+        StatusCode,
+    };
+    use hyper::Body;
+    use rand::Rng;
+    use tower::{BoxError, Service, ServiceBuilder, ServiceExt};
+
+    use crate::{async_session::MemoryStore, SessionHandle, SessionLayer};
+
+    #[tokio::test]
+    async fn sets_session_cookie() {
+        let secret = rand::thread_rng().gen::<[u8; 64]>();
+        let store = MemoryStore::new();
+        let session_layer = SessionLayer::new(store, &secret);
+        let mut service = ServiceBuilder::new().layer(session_layer).service_fn(echo);
+
+        let request = Request::get("/").body(Body::empty()).unwrap();
+
+        let res = service.ready().await.unwrap().call(request).await.unwrap();
+        assert_eq!(res.status(), StatusCode::OK);
+
+        assert!(res
+            .headers()
+            .get(SET_COOKIE)
+            .unwrap()
+            .to_str()
+            .unwrap()
+            .starts_with("axum.sid="))
+    }
+
+    #[tokio::test]
+    async fn unsaved_changed_false_does_not_set_cookie() {
+        let secret = rand::thread_rng().gen::<[u8; 64]>();
+        let store = MemoryStore::new();
+        let session_layer = SessionLayer::new(store, &secret).with_save_unchanged(false);
+        let mut service = ServiceBuilder::new().layer(session_layer).service_fn(echo);
+
+        let request = Request::get("/").body(Body::empty()).unwrap();
+
+        let res = service.ready().await.unwrap().call(request).await.unwrap();
+        assert_eq!(res.status(), StatusCode::OK);
+
+        let session_cookie = res
+            .headers()
+            .get(SET_COOKIE)
+            .unwrap()
+            .to_str()
+            .unwrap()
+            .to_string();
+        let mut request = Request::get("/").body(Body::empty()).unwrap();
+        request
+            .headers_mut()
+            .insert(COOKIE, session_cookie.parse().unwrap());
+        let res = service.ready().await.unwrap().call(request).await.unwrap();
+        assert!(res.headers().get(SET_COOKIE).is_none());
+    }
+
+    #[tokio::test]
+    async fn uses_valid_session() {
+        let secret = rand::thread_rng().gen::<[u8; 64]>();
+        let store = MemoryStore::new();
+        let session_layer = SessionLayer::new(store, &secret);
+        let mut service = ServiceBuilder::new().layer(session_layer).service_fn(echo);
+
+        let request = Request::get("/").body(Body::empty()).unwrap();
+
+        let res = service.ready().await.unwrap().call(request).await.unwrap();
+        assert_eq!(res.status(), StatusCode::OK);
+
+        let session_cookie = res.headers().get(SET_COOKIE);
+        let mut request = Request::get("/").body(Body::empty()).unwrap();
+        request
+            .headers_mut()
+            .insert(COOKIE, session_cookie.unwrap().to_owned());
+        let res = service.ready().await.unwrap().call(request).await.unwrap();
+        assert_eq!(res.headers().get(SET_COOKIE), session_cookie);
```

Comment:
I fully understand that the `cookie_value.is_none()` condition is necessary; I guess I've just never tried running with `save_unchanged = false` :smile: .

Currently I'm trying to wrap my head around why the `Ok(None)` part is necessary in the middleware, and also this corresponding test case. As far as I understand: even though if the request has the correct cookie, and nothing has effectively changed, we always send back the `Set-Cookie` header. 
Can you point me to the right direction why is this necessary? I might be wrong, I'm not an expert.

Reply:
If I understand the semantic of "save unchanged" (as ported from the Tide middleware) we should always set the cookie if this is true. (Let me know if this makes sense, I might be misunderstanding the Tide logic).

What I realized is that storing the session will only return a cookie value the first time (this could be a bug, I'm not sure; I wrote this code on a plane without WiFi so I was doing my best to look through the async-session code but didn't have access to Tide directly), so if we don't add this to the `None` case then we don't actually set cookie.

There might be a better way of doing this or I might just be plain wrong about the intended semantic. 😁

Thank you for taking a look! 🙇

Reply:
I think tide [only sends the cookie if `store_session` returns `Some`](https://github.com/http-rs/tide/blob/main/src/sessions/middleware.rs#L114-L124). 

I'm also a little confused now, tried to put unreachable to this `Ok(None)` branch temporarily, and I literally cannot hit it with my "real" application, or curl. I guess this needs a little more investigation on my part. Additionally, it might be worth checking what actix-session does, just for reference.

Reply:
Could it be something about `MemoryStore` specifically?

I seem to recall hitting that branch when I was setting one of the examples but I don't remember which now.

Reply:
I suspect it's MemoryStore, but I'll need to look into it more thoroughly. It calls [`reset_data_changed`](https://github.com/http-rs/async-session/blob/0f825c3827a0bf420b8d0ce7a011f9b7a9c3671c/src/memory_store.rs#L55) inside `store_session`, unlike `async-redis-session` or `async-sqlx-session` .